modules/http4s/src/smithy4s/http4s/internals/SmithyHttp4sClientEndpoint.scala

```diff
@@ -143,29 +191,80 @@ private[smithy4s] class SmithyHttp4sClientEndpointImpl[F[_], Op[_, _, _, _, _],
       oneOf: SchemaAlt[E, ErrorType],
       response: Response[F]
   ): F[O] = {
+    tryProcessError(errorable, oneOf, response).rethrow
+      .map(errorable.unliftError)
+      .flatMap(effect.raiseError)
+  }
+
+  private def tryProcessError[ErrorType](
+      errorable: Errorable[E],
+      oneOf: SchemaAlt[E, ErrorType],
+      response: Response[F]
+  ): F[Either[MetadataError, E]] = {
     val schema = oneOf.instance
     val errorMetadataDecoder = Metadata.PartialDecoder.fromSchema(schema)
     implicit val errorCodec = entityCompiler.compilePartialEntityDecoder(schema)
     decodeResponse[ErrorType](response, errorMetadataDecoder)
-      .map(oneOf.inject)
-      .map(errorable.unliftError)
-      .flatMap(effect.raiseError)
+      .map(_.map(oneOf.inject))
   }
 
   private def decodeResponse[T](
       response: Response[F],
       metadataDecoder: Metadata.PartialDecoder[T]
-  )(implicit entityDecoder: EntityDecoder[F, BodyPartial[T]]): F[T] = {
+  )(implicit
+      entityDecoder: EntityDecoder[F, BodyPartial[T]]
+  ): F[Either[MetadataError, T]] = {
     val headers = getHeaders(response)
     val metadata = Metadata(headers = headers)
     metadataDecoder.total match {
       case Some(totalDecoder) =>
-        totalDecoder.decode(metadata).liftTo[F]
+        totalDecoder.decode(metadata).pure[F]
       case None =>
         for {
-          metadataPartial <- metadataDecoder.decode(metadata).liftTo[F]
+          metadataPartial <- metadataDecoder.decode(metadata).pure[F]
           bodyPartial <- response.as[BodyPartial[T]]
-        } yield metadataPartial.combine(bodyPartial)
+        } yield metadataPartial.map(_.combine(bodyPartial))
     }
   }
 }
+
+private final class ErrorAltPicker[E](alts: Vector[SchemaAlt[E, _]]) {
```

Comment:
This doesn't seem to be depending on any http4s-specific construct. Let's move it to core (in package `smithy4s.http` for instance)

Reply:
I have a dependency on cats for `===`, should I just replace that to `==`?

Reply:
used `==`